### DISCUSSION TITLE
Bluetooth gamepad controller

### DISCUSSION THREAD

**Discussion by DanMelbourne** [dirkwhoffmann/vAmiga#826] [Q&A]
Can anyone recommend an inexpensive Bluetooth controller that would work with Vamiga? I was thinking of something like the 8bitdo Pocket-Sized Mini Controller Micro Bluetooth Gamepad (https://www.amazon.com.au/8Bitdo-Pocket-Sized-Controller-Bluetooth-Gamepad/dp/B0CDG2HKBF?th=1) but can't find any information about compatibility of Bluetooth controllers with Vamiga.

**Comment by dirkwhoffmann** [comment]
I can't tell if vAmiga will work with that particular device, but my experience with such cheap devices isn't that good. I'm also unsure about latency issues (I would expect the latency of wireless devices much higher than the latency of, e.g., ordinary USB devices, but I never did any measurements). I can recommend the USB version of the [Competion Pro ](https://www.amazon.de/SPEEDLINK-SL-650212-BKRD-Competition-EXTRA-Joystick/dp/B07K9N36GG/ref=sr_1_1?__mk_de_DE=ÅMÅŽÕÑ&dib=eyJ2IjoiMSJ9.3vPnl1FSLb0QadFQX95XsbBsxqg6t37aN8EQaHTEFuxvZ7qcnfll5bzvJRELsUMlWtn2c-g8aDI0Zh4PHPylwpgOXu1vFsmeTnnlDhiV-DAY_sKEu3BbU05uA2mAhtpaNJBetsm5K0e1lmi0poQybiQgV0-PGQXxCiwW61e10dOc0MyBiJ1TtrLkC-nCjhEPj2R0al6kR43bsAoQBoc44iMp5p4NZomQXuN7O2-6BsQ.HRxLko7-AzNu_D3mXZe9l7nr31fKCs08oAP4JvvN7HA&dib_tag=se&keywords=speedlink+competition+pro&qid=1716540560&sr=8-1) (which isn't wireless though):

**Comment by DanMelbourne** [comment]
Thanks for the quick reply Dirk! I suppose I could buy them and return them to Amazon if they don't work. That's the nice thing about Amazon... very liberal return policy!

**Comment by DanMelbourne** [comment]
So I ordered these gamepads and they work fine while plugged in using USB but I couldn't get them working in Bluetooth mode at all – vAmiga only "saw" them as external devices that could be assigned to a port if plugged in over USB.  I'm actually returning them to Amazon because the gamepad tends to transmit two button presses rather than one on the four way rocker, which makes it hard to play Super Twintris precisely! But good to know now that 8bitdo gamepads do work with vAmiga - I wonder if all USB gamepads generally would.

Are there any Bluetooth controllers at all known to work with vAmiga?

**Comment by mithrendal** [comment]
@DanMelbourne 

ok lets check on macbook air m1

1. connected an XBox wireless controller via bluetooth

<img width="804" alt="image" src="https://github.com/dirkwhoffmann/vAmiga/assets/17108995/3ba00517-3adc-4f27-875e-f01d2b8421e2">


2. start vAmiga and the controller was selectable straight from the port2 drop down list

3. started Amiga Test Kit proves all directions are working fine


<img width="861" alt="image" src="https://github.com/dirkwhoffmann/vAmiga/assets/17108995/d7b18d64-0852-4827-9247-2bd3b098c21d">

**Comment by dirkwhoffmann** [comment]
`DeviceDatabase.swift` keeps the mapping schemes of all known devices: 

```Swift
       // Setup the lookup table for all known devices
        known =
            ["4":
                ["1": [:]],                     // aJoy Retro Adapter
             "13":
                ["0": ["L": "1", "R": "3", "H": "4"]], // Nimbus+
             "121":
                ["17": ["R": "1"],              // iNNEXT Retro (SNES)
                 "6354": [:]],                  // Mayflash Magic-NS 1.2
             "1035":
                ["25907": [:]],                 // Competition Pro SL-6602
             "1118":
                ["2835": ["H": "1"],            // XBox Carbon Black
                 "746": ["H": "3"]],            // XBox One Wired Controller
             "1133":
                ["49250": [:]],                 // Logitech Mouse
             "1155":
                ["36869": [:]],                 // RetroFun! Joystick Adapter
             "1356":
                ["616": [:],                    // Sony DualShock 3
                 "1476": [:],                   // Sony DualShock 4
                 "2508": [:],                   // Sony Dualshock 4 (2nd Gen)
                 "3302": [:]],                  // Sony DualSense
             "1848":
                ["8727": [:]],                  // Competition Pro SL-650212
             "3853":
                ["193": ["R": "1"]],            // HORIPAD for Nintendo Switch
             "7257":
                ["36": [:]]                     // The C64 Joystick
            ]
```

Among them, the XBox controllers, the Sony controllers, and the Nimbus are wireless.

**Reply by DanMelbourne** [reply]
Oh awesome! Thank you! I also found that the Bluetooth 'NES controllers' I have (for Nintendo Switch -- sold by Nintendo) also work fine with vAmiga which is super cool.

**Comment by DanMelbourne** [comment]
Just confirming that the Nintendo Switch "NES Controller"s work like a dream in vAmiga! $AUD79 for a pair of them. They show in your Bluetooth list separately as NES Controller (L) and NES Controller (R), as well as in vAmiga. 

Perfect for reliving my glory years with my best buddy playing Super Twintris!! 

<img width="312" alt="Screenshot 2024-05-30 at 9 18 20 PM" src="https://github.com/dirkwhoffmann/vAmiga/assets/32061007/c73b402c-24b5-4148-92b0-51632e4ba07f">

<img width="272" alt="Screenshot 2024-05-30 at 9 32 02 PM" src="https://github.com/dirkwhoffmann/vAmiga/assets/32061007/0f595751-34e6-4fee-a0d7-471233d5d987">

I did find vAmiga can be a bit finicky about detecting both of them -- sometimes it only detects one of them. But quitting and restarting the app usually fixes it.
